### ISSUE TITLE
Establish simple web page to replace apps.hydroshare.org

### ISSUE THREAD

**Issue by dtarb** [hydroshare/hydroshare#2301]
Develop/establish simple web page with buttons for launching CUAHSI approved apps to replace apps.hydroshare.org.  I suggest this web page be hosted at www.hydroshare.org/apps and that this have a layout similar to apps.hydroshare.org (but consistent with the www.hydroshare.org style) with a button for launching each app from its icon and an “i” button that you can hover over to learn more information and possibly click through to get to the help page.

- This page should be automatically generated from resources that are “Web App Resource” type and on a list approved by CUAHSI.  CUAHSI should have a simple way to manage resources on this list, perhaps as a CSV file saved somewhere (e.g. in a HS resource they have access to)
- This page should not require login (though apps.hydroshare.org currently does)
- Launching an app on this page should open the URL in the Home Page URL entry in the web app resource.

This idea was discussed with @danames @zhiyuli and @Castronova in an email thread in May 2017.  The motivation is to avoid having to write a tethys app to implement a simple button for every third party app that we want to support (which is the case now with Jupyter, SWATshare and CyberGIS apps.  This also provides an easy way for CUAHSI to manage approved apps.

There are two related actions

1. The same list of CUAHSI approved apps should be connected to Open With functionality - see #1161.
2. The apps.hydroshare.org page should be redirected to this new page so as not to break links to apps.hydroshare.org.  

One implementation approach is to use a RTF Mezzanine page.  I have experimented with this approach a bit on riverdev.hydroshare.org, where (as admin) I did the following

- Created a new rich text format page https://riverdev.hydroshare.org/app-page/ and using HTML constructed icons and links to launch apps
- Redirected the apps top navigation link to this page.

If this approach was taken, code would need to be written to update this page based on the approved app list.

**Comment by aphelionz** [comment]
@dtarb @hyi @pkdash 

I started working on this and got decently far. I have a basic layout but I'm missing key pieces of metadata and am not sure where to find it. Namely I need a description, the URL of the web app, and a thumbnail or icon for it.

I tried both `ToolResource.objects.all()[0].metadata.homepage_url` and `ToolResource.objects.all()[0].metadata.toolmetadata.homepage_url` for example and can't really find much.

Are the data I need missing or am I looking in the wrong place?

**Comment by pkdash** [comment]
@aphelionz 

To get the url of the app home page:
```ToolResource.objects.all()[0].metadata.app_home_page_url```

Similarly you can see all the metadata properties that you can access for an instance of ToolResource here:
https://github.com/hydroshare/hydroshare/blob/develop/hs_tools_resource/models.py#L292

**Comment by aphelionz** [comment]
Thank you sir!

**Comment by aphelionz** [comment]
Here's a glimpse into what I've done so far...

![image](https://user-images.githubusercontent.com/106148/30516304-fb1e9550-9b08-11e7-935a-e433db249d3f.png)

This is a simple listing of every "ToolResource" aka "Web App" in our database that has an icon and a URL. My thoughts right now are:

1. We obviously need to style this properly so that it's consistent with the rest of the pages. Any insight into that would be helpful as I'm not a designer - right now it's purely "functional." However I don't know that we should do the "Pinterest style" layout from apps as that's a JavaScript enhancement and I don't think we need to add that complexity to this page.
2. I need to add two pieces of metadata to the model - `description` and `approved.` The former so that we can add the info popup for each app and the latter so that we can filter them more granularly.
3. Do we need a `category` as well for the apps? Any other pieces of metadata that we can think of?

**Comment by aphelionz** [comment]
Oh and also @dtarb or whomever you delegate to. What should the paragraph at the top say to replace "Quick block of text describing the apps, our methodology, etc..."

**Comment by dtarb** [comment]
I suggest the following lead in text:

HydroShare apps allow you to visualize, analyze, and work with resources (data and models) in HydroShare.  Apps are hosted on separate web servers from the HydroShare website (www.hydroshare.org) and access HydroShare resources using web services via the REST applications programmers interface (API). Anyone can write an app an then create a ["Web App" resource](https://help.hydroshare.org/hydroshare-resource-types/web-app/) that holds the configuration information for launching the App from HydroShare.  This page lists [CUAHSI approved web apps](https://help.hydroshare.org/apps/) that are supported as part of HydroShare. 

I suggest go with this but be ready to refine once we demo and get feedback from the team.

The way I have written this does commit us to maintaining documentation on https://help.hydroshare.org/apps/ for approved apps.

Note that the multiple SWAT apps that we had at one point have been consolidated down to One.  See apps.hydroshare.org.  

If we are to have description and category, ideally these should be part of the Web App resource and drawn from that so they can be maintained by the app developer.  @danames @zhiyuli please consider this in the Web App resource revision with #1610.  We could use the Web App resource Abstract as the Description.  I suggest do this now, so as not to depend on the App resource revisions, with coding so it is easy to change to a separate description field if we decide that what appears in App Hover over should be different than the Web App resource abstract. 

I do not think we are at a point where we need a category, but we may in the future.  This should be a controlled vocabulary when implemented.  I do not have a firm idea of what categories would be needed.

**Comment by aphelionz** [comment]
@dtarb 

Thanks, I updated the text and will make a commit soon. I can use the abstract, I think. No need to create a separate description field and we can leave the category out for now.

However, I do think we should add an "approved" flag to the `ToolResource` model. Can that be part of #1610 or should I do it as part of this effort?

**Comment by dtarb** [comment]
The policy that approval of a web should be by CUAHSI is set.  I do not have a good opinion as to what is the best tactic for implementation, but if it is via an approved flag on the resource, then setting that flag will have to somehow be restricted to the list of allowed approvers, not the owners or editors of the resource as is the default for other resource metadata.  I dont really mind which issue of the 4 related #1161, #1609, #1610 and this it is done with.  These are all linked.

**Comment by aphelionz** [comment]
@dtarb I think we're just gonna keep the list of apps in a python array for now. It's simple, allows us to use the Django query set, and can be easily modified to add an app, which won't happen very often.

Here's some more UI stuff: http://recordit.co/tVQBBA4SND Note that we'll probably have to enforce some limitations on the description... or maybe truncate with a link to the resource page?

**Comment by aphelionz** [comment]
https://playground.hydroshare.org/apps/

**Comment by dtarb** [comment]
General layout seems reasonable.  On a wide screen icons are very widely spaced, consider either wrapping or grouping.  Find a way to deal with explanatory text that does not fit.  E.g. Jupyterhub NCSA text does not fit.  Cull down to only the following approved Apps
- Swatshare (only the one without qualifiers)
- HydroShare GIS
- JupyterHub NCSA
- CUAHSI Data Series Viewer
- National Water Model Forecast Viewer
- Data Rods Explorer
Some of these may not even be up to par, but lets go from here and App champions can champion for others.

**Comment by aphelionz** [comment]
@dtarb Another revision is up at https://playground.hydroshare.org/apps/

The only app that I couldn't add was the Data Rods Explorer. I couldn't find a webapp resource so we'll need to make a resource for that in HydroShare beta and prod when we finally test and deploy, respectively. 

The apps that are currently on apps.hydroshare that would require championship are:

* Snow Inspector
* Cybergis TauDEM
* EPA Waters
* USGS Gaugeviewer

I welcome your additional feedback.

**Comment by danames** [comment]
Please add EPA Waters and Snow Inspector. I use them often in demonstrations and also in my classes. 

Also, I still think that the list of these apps should be defined in a resource, not hard coded in code. Maybe you've already made that change, but I think it is a no-brainer to not hard-code the apps list.

**Comment by aphelionz** [comment]
@dtarb @danames This will be on park momentarily. Please see Pull request #2322 for instructions on how to test.

**Comment by dtarb** [comment]
I tested following the protocol in #2322 and found that the test case logged in as an Admin works. 

However not logged in as admin I noticed

- Trying the test case for users without admin access, I am not able to see the 7 fields mentioned.  I had to log in as admin to get into edit mode to see the 7 fields.  These 7 fields are only visible to someone with edit permissions on the web app instance resource.

- The manage access button was missing for the existing web app instance resources I looked at, but was present for a new one I created by copying.  I do not think this should be the case.  Web App Instances should have access management functionality the same as any other resource.  (I noticed this because I wanted to set access to I could get into edit mode to look at the 7 fields).

- The "Show on Open With" checkbox should be visible in view and edit mode.  Because if I have view access to a web app instance (that is not admin approved) and I want it on open with, I should still be able to set it.  

- The "Show on Open With" functionality is not working.  I created a copy of a web app resource.  I verified that this web app was not approved.  I set the Show On Open With, but this does not show up on my show on open with button.
